mijia/src/lib.rs

```diff
@@ -17,85 +18,127 @@ const CONNECTION_INTERVAL_CHARACTERISTIC_PATH: &str = "/service0021/char0045";
                 })
-            } else {
-                None
             }
-        })
-        .collect();
-    Ok(sensors)
+            Some(BluetoothEvent::Connected {
+                object_path,
+                connected: false,
+            }) => Some(MijiaEvent::Disconnected {
+                id: DeviceId { object_path },
+            }),
+            _ => None,
+        }
+    }
 }
 
-pub async fn start_notify_sensor(
-    bt_session: &MijiaSession,
-    device_path: &str,
-) -> Result<(), anyhow::Error> {
-    let temp_humidity_path = device_path.to_string() + SENSOR_READING_CHARACTERISTIC_PATH;
-    let temp_humidity = dbus::nonblock::Proxy::new(
-        "org.bluez",
-        temp_humidity_path,
-        DBUS_METHOD_CALL_TIMEOUT,
-        bt_session.connection.clone(),
-    );
-    temp_humidity.start_notify().await?;
+/// A wrapper around a bluetooth session which adds some methods for dealing with Mijia sensors.
+/// The underlying bluetooth session may still be accessed.
+pub struct MijiaSession {
+    pub session: BluetoothSession,
+}
+
+impl MijiaSession {
+    /// Returns a tuple of (join handle, Self).
+    /// If the join handle ever completes then you're in trouble and should
+    /// probably restart the process.
+    pub async fn new(
+    ) -> Result<(impl Future<Output = Result<(), anyhow::Error>>, Self), anyhow::Error> {
+        let (handle, session) = BluetoothSession::new().await?;
+        Ok((handle, MijiaSession { session }))
+    }
+
+    /// Get a list of all Mijia sensors which have currently been discovered.
+    pub async fn get_sensors(&self) -> Result<Vec<SensorProps>, anyhow::Error> {
+        let devices = self.session.get_devices().await?;
+
+        let sensors = devices
+            .into_iter()
+            .filter_map(|device| {
+                if device.service_data.contains_key(MIJIA_SERVICE_DATA_UUID) {
+                    Some(SensorProps {
+                        id: device.id,
+                        mac_address: device.mac_address,
+                    })
+                } else {
+                    None
+                }
+            })
+            .collect();
+        Ok(sensors)
+    }
+
+    /// Assuming that the given device ID refers to a Mijia sensor device and that it has already
+    /// been connected, subscribe to notifications of temperature/humidity readings, and reduce the
+    /// connection interval to save power.
+    ///
+    /// Notifications will be delivered as events by `MijiaSession::event_stream()`.
+    pub async fn start_notify_sensor(&self, id: &DeviceId) -> Result<(), anyhow::Error> {
+        let temp_humidity_path = id.object_path.to_string() + SENSOR_READING_CHARACTERISTIC_PATH;
+        let temp_humidity = dbus::nonblock::Proxy::new(
+            "org.bluez",
+            temp_humidity_path,
+            DBUS_METHOD_CALL_TIMEOUT,
+            self.session.connection.clone(),
+        );
+        temp_humidity.start_notify().await?;
```

Comment:
your bluetooth.rs split is such a good idea that I'm suddenly trying to think of ways to move things like this into bluetooth.rs (definitely not in this PR).